flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/CommonPythonCalc.scala

```
@@ -0,0 +1,72 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.table.plan.nodes
+
+import org.apache.calcite.rex.{RexCall, RexInputRef, RexNode}
+import org.apache.flink.table.functions.FunctionLanguage
+import org.apache.flink.table.functions.python.{PythonFunction, PythonFunctionInfo, SimplePythonFunction}
+import org.apache.flink.table.functions.utils.ScalarSqlFunction
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+
+trait CommonPythonCalc {
+
+  private[flink] def extractPythonScalarFunctionInfos(
+      rexCalls: Array[RexCall]): (Array[Int], Array[PythonFunctionInfo]) = {
+    // using LinkedHashMap to keep the insert order
+    val inputNodes = new mutable.LinkedHashMap[RexNode, Integer]()
+    val pythonFunctionInfos = rexCalls.map(createPythonScalarFunctionInfo(_, inputNodes))
+
+    val udfInputOffsets = inputNodes.toArray.sortBy(_._2).map(_._1).map {
+      case inputRef: RexInputRef => inputRef.getIndex
+    }
+    (udfInputOffsets, pythonFunctionInfos)
+  }
+
+  private[flink] def createPythonScalarFunctionInfo(
+      rexCall: RexCall,
+      inputNodes: mutable.Map[RexNode, Integer]): PythonFunctionInfo = rexCall.getOperator match {
+    case sfc: ScalarSqlFunction if sfc.getScalarFunction.getLanguage == FunctionLanguage.PYTHON =>
+      val inputs = new mutable.ArrayBuffer[AnyRef]()
+      rexCall.getOperands.foreach {
+        case pythonRexCall: RexCall if pythonRexCall.getOperator.asInstanceOf[ScalarSqlFunction]
+          .getScalarFunction.getLanguage == FunctionLanguage.PYTHON =>
+          // Continuous Python UDFs can be chained together
+          val argPythonInfo = createPythonScalarFunctionInfo(pythonRexCall, inputNodes)
+          inputs.append(argPythonInfo)
+
+        case argNode: RexNode =>
```

Should we add exceptions and meaningful exception messages here if the RexNode is a RexLiteral? We don't support UDFs with literals now.

For example, we can do the check and throw the exception in the DataStreamPythonCalcRule, thus we can check it through the plan test. Or can we throw the exception earlier?

---

Good catch! Python UDF with constant parameters will be supported in FLINK-14208. Will throw an exception here.